test/pullrequest/e2e_test.go

```diff
@@ -0,0 +1,200 @@
+)
+
+func TestPullRequest(t *testing.T) {
+	tmpdir, err := ioutil.TempDir("", t.Name())
+	if err != nil {
+		t.Fatalf("could not create test temp dir: %v", err)
+	}
+	t.Logf("Test dir: %s", tmpdir)
+	defer os.RemoveAll(tmpdir)
+
+	updateDeps(t)
+
+	for _, tc := range []struct {
+		name string
+		url  string
+
+		// Location of fake SCM data to use for proxy.
+		scmData string
+
+		// Auth token to use.
+		token string
+		// If auth token present, fmt template for Authorization header to check.
+		authHeader string
+	}{
+		{
+			name: "github-default",
+			// Must be http so we can intercept the request in the proxy.
+			url:        "http://github.com/wlynch/test/pulls/1",
+			scmData:    filepath.Join("testdata", "scm", "api.github.com"),
+			token:      envToken("GITHUB_TOKEN"),
+			authHeader: "Bearer %s",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			dir := filepath.Join(tmpdir, tc.name)
+			if err := os.Mkdir(dir, 0755); err != nil {
+				t.Fatalf("os.Mkdir: %v", err)
+			}
+
+			// If provided, set the AUTH_TOKEN environment variable that the
+			// pullrequest binary expects.
+			if tc.token != "" {
+				os.Setenv("AUTH_TOKEN", tc.token)
+				defer os.Unsetenv("AUTH_TOKEN")
+			}
+
+			if *proxy {
+				var header http.Header
+				if tc.token != "" {
+					header = http.Header{
+						"Authorization": []string{fmt.Sprintf(tc.authHeader, tc.token)},
+					}
+				}
+				startProxy(t, tc.scmData, header)
+			}
+
+			t.Log("Environment vars:", os.Environ())
+
+			// Download PR data to local dir.
+			runCommand(t, "download", dir, tc.url)
+			// Diff downloaded contents with golden copies.
+			golden := filepath.Join("testdata", "golden")
+			if diff, err := exec.Command("diff", "-r", golden, dir).CombinedOutput(); err != nil {
+				t.Error(string(diff))
+			}
+
+			// Modify PR resource contents.
+			modifyFiles(t, dir)
+
+			// Upload new state back to server.
+			runCommand(t, "upload", dir, tc.url)
+		})
+	}
+}
+
+// Stand up fake proxy server. This should not be called if you want to talk to
+// real SCM providers, since this will intercept all HTTP traffic.
+func startProxy(t *testing.T, dataPath string, header http.Header) {
+	t.Helper()
+
+	h := &handler{
+		Data:   dataPath,
+		Header: header,
+	}
+
+	// Start server on any open port.
+	l, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("net.Listen: %v", err)
+	}
+	t.Logf("starting HTTP proxy on %s", l.Addr().String())
+	go func() {
+		if err := http.Serve(l, h); err != nil {
+			fmt.Println("http.Serve:", err)
+		}
+	}()
+	os.Setenv("http_proxy", l.Addr().String())
+}
+
+// updateDeps makes sure command is installed locally to resolve any
+// dependencies. We need to do this first since the run might try to
+// pull in new dependencies through the proxy.
+func updateDeps(t *testing.T) {
+	t.Helper()
+	install := exec.Command("go", "get", "-u", pkg)
```

Comment:
This can't be done in the `CI` (because of `-mod=vendor`). Our tests should do as little "network" as possible. A `exec.Command("go", "build", pkg)` should work though :wink:

Reply:
All the dependencies are already there (in the `vendor` folder) and even if we didn't have the `vendor` folder, doing a `go build` (without `-mod=vendor`) would get those dependencies. Doing a `go get -u` here scares me to death as I don't have the guarantee it will use the code that is in the PR (for `pullrequest-init`)

Reply:
Good call. Done.